src/Database/Concerns/HasRelationships.php

```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
-
-        $primaryKey = $primaryKey ?: $this->getForeignKey();
-
-        $foreignKey = $foreignKey ?: $instance->getForeignKey();
-
-        if (is_null($table)) {
-            $table = $this->joiningTable($related);
+    /**
+     * {@inheritDoc}
+     */
+    protected function newHasMany(Builder $query, Model $parent, $foreignKey, $localKey)
+    {
+        $relation = new HasMany($query, $parent, $foreignKey, $localKey);
+        $caller = $this->getRelationCaller();
+        if (!is_null($caller)) {
+            $relation->setRelationName($caller);
         }
-
-        return new BelongsToMany(
-            $instance->newQuery(),
-            $this,
-            $table,
-            $primaryKey,
-            $foreignKey,
-            $parentKey ?: $this->getKeyName(),
-            $relatedKey ?: $instance->getKeyName(),
-            $relationName
-        );
+        return $relation;
     }
 
     /**
-     * Define a polymorphic many-to-many relationship.
-     * This code is almost a duplicate of Eloquent but uses a Storm relation class.
-     * @return \Winter\Storm\Database\Relations\MorphToMany
+     * {@inheritDoc}
      */
-    public function morphToMany($related, $name, $table = null, $primaryKey = null, $foreignKey = null, $parentKey = null, $relatedKey = null, $inverse = false, $relationName = null)
+    protected function newHasManyThrough(Builder $query, Model $farParent, Model $throughParent, $firstKey, $secondKey, $localKey, $secondLocalKey)
     {
-        if (is_null($relationName)) {
-            $relationName = $this->getRelationCaller();
+        $relation = new HasManyThrough($query, $farParent, $throughParent, $firstKey, $secondKey, $localKey, $secondLocalKey);
+        $caller = $this->getRelationCaller();
+        if (!is_null($caller)) {
+            $relation->setRelationName($caller);
         }
-
-        $instance = $this->newRelatedInstance($related);
-
-        $primaryKey = $primaryKey ?: $name.'_id';
-
-        $foreignKey = $foreignKey ?: $instance->getForeignKey();
-
-        $table = $table ?: Str::plural($name);
-
-        return new MorphToMany(
-            $instance->newQuery(),
-            $this,
-            $name,
-            $table,
-            $primaryKey,
-            $foreignKey,
-            $parentKey ?: $this->getKeyName(),
-            $relatedKey ?: $instance->getKeyName(),
-            $relationName,
-            $inverse
-        );
+        return $relation;
     }
 
     /**
-     * Define a polymorphic many-to-many inverse relationship.
-     * This code is almost a duplicate of Eloquent but uses a Storm relation class.
-     * @return \Winter\Storm\Database\Relations\MorphToMany
+     * {@inheritDoc}
      */
-    public function morphedByMany($related, $name, $table = null, $primaryKey = null, $foreignKey = null, $parentKey = null, $relatedKey = null, $relationName = null)
+    protected function newMorphMany(Builder $query, Model $parent, $type, $id, $localKey)
```

Comment:
Are these `new$relationType` methods in the same order as the old ones? Just wondering in case we could get a cleaner diff for them. Also is there any concerns with the public methods being removed?

Reply:
There's two changes happening here:

1. The original public methods are no longer in our `HasRelationships` definition. These are now provided by Laravel core.
2. The protected `new$relationType` methods are what we extend from the original Laravel definitions, mainly to provide a relation name and some small adjustments for our own needs. I believe these were *not* in the original file, so that's the reason for the messy diff.